File: CS 471 - Prog Lang Structure/Programming #4/gaussian.f90
!
! Name        : Keller Sedillo-Garrido
! Date        : September 28, 2022
! Description : This program will perform Gaussian Elimination and Backward Subitution
!             : with a NxN+1 Matrix
! Input       : The size of the matrix
! Output      : A solved matrix with solutions.
!
program gaussian
  implicit none
  
  ! Set Vars
  integer N, i, j, k                     ! Integers 
  real ratio                             ! Real Nums
  real, dimension(:,:), allocatable :: A ! 2-D Array for Matrix
  real, dimension(:), allocatable :: x   ! 1-D Array for Solutions
  character(100) :: input                ! User Input

  ! Read in Command Arg
  call GET_COMMAND_ARGUMENT(1, input) ! Get input from users
  read(input,*)N                      ! Set input as N

  !Allocate Space
  ALLOCATE(A(N ,N+1)) ! Allocate for Matrix
  ALLOCATE(x(N))      ! Allocate for Solutions

  !Randomizes Number
  call random_number(A)             ! Give array random numbers
  do i = 1, N                       ! Loop through row
    do j = 1, N+1                   ! Loop through col
      A(i,j) = 1 + FLOOR(10*A(i,j)) ! Set Whole numbers
    end do                          ! End for
  end do                            ! End for 

  
  !Gaussian Elimination
  do i = 1, N                             ! Loop through col
    if ( A(i,i) == 0 ) then               ! Check if diag = 0
      A(i,i) = 1                          ! Set it to 1
    end if                                ! End if
    do j = i+1, N                         ! Loop through row
      ratio = A(j,i)/A(i,i)               ! Set ratio
      do k = 1, N+1                       ! Loop through row
        A(j,k) = A(j, k) - ratio * A(i,k) ! Update elements
      end do                              ! End for
    end do                                ! End for
  end do                                  ! End for
  
  !Backward Sub.
  x(N) = A(N,N+1)/A(N,N)        ! Set init. Solution
  do i = N-1, 1, -1             ! Loop backwards from Solution array
    x(i) = A(i,N)               ! Set Elements
    do j = i+1, N               ! Loop through col
      x(i) = x(i) - A(i,j)*x(j) ! Update Elements
    end do                      ! End for
    x(i) = x(i)/A(i,i)          ! Update solution
  end do                        ! End for
  
end program gaussian

File: CS 471 - Prog Lang Structure/Programming #4/gaussian.py
#
# Name        : Keller Sedillo-Garrido
# Date        : September 28, 2022
# Description : This program will perform Gaussian Elimination and Backward Subitution
#             : with a NxN+1 Matrix
# Input       : The size of the matrix
# Output      : A solved matrix with solutions.
#

#Libaries
import random
import sys
from array import *

#Set Vars
N = int(sys.argv[1]) # Get size of Matrix
A = []               # Create array for Matrix
x = []               # Create array for Solutions

#Create Matrix with random numbers
for i in range(N):                                             # Loop through the rows
    col = []                                                   # Create Columns  
    for j in range(N+1):                                       # Loop through the columns
        col.append(random.choice([1, 2, 3, 4, 5, 6, 7, 8, 9])) # Append a random number
    A.append(col)                                              # Add to array

#Gaussian Elimination
for i in range(N):                              # Loop Through Rows
    for j in range(i+1, N):                     # Loop Through Columns
        ratio = A[j][i]/A[i][i]                 # Get Ratio
        for k in range(N+1):                    # Loop through the rows
            A[j][k] = A[j][k] - ratio * A[i][k] # Update the elements

#Set the elements in the arrays in zero.
for i in range(N): 
    x.append(0)

#Set inital Solution
x[N-1] = A[N-1][N]/A[N-1][N-1]

#Backward Sub.
for i in range(N-2, -1, -1):       # Loop backward from end of array
    x[i] = A[i][N]                 # Set elemetns
    for j in range(i+1, N):        # Loop through columns
        x[i] = x[i] - A[i][j]*x[j] # Update elements
    x[i]=x[i]/A[i][i]              # Update element with solution
